Feature - Added a hook system into the baseline scripts & scans #4925

rezen opened the pull request:
As per the feature #4925 created a hook system to enable users to augment or add to the baseline/full/api scans. By default, the scripts will look for `~/.zap_hooks.py` but you can supply a `--hook=path-to-custom/hooks.py`

For each hook you want to plugin in to, you create a method with the same name as the hook and then there you can add to or change what zap is doing. Every hook that has the `zap_` prefix will pass the zap api object which you can interact with. Any hook defined in `zap_common.py` that uses the `@hook` decorator, you are intercepting the arguments that are passed to the method (the hook is the same name as the method) and you can modify the arguments. Every hook that ends with `_wrap` allows you to wrap the results/return value of the method of the same name.

**Example** 
```python
def start_docker_zap_wrap(container_id):
  return container_id

def zap_started(zap, target):
  print(zap)

def zap_access_target(zap, target):
  print("Access target %s" % target)

def zap_spider(zap, target):
  pass

def zap_get_alerts(zap, baseurl, blacklist, out_of_scope):
  blacklist.append('20001')
  return zap, baseurl, blacklist, out_of_scope

def zap_get_alerts_wrap(results):
  return results

def zap_spider_wrap(zap=None):
  print("Post", zap)
  raise Exception("test")

def zap_pre_shutdown(zap=None):
  pass
```

psiinon commented:
This pull request **introduces 3 alerts** when merging a401aa12bfc02e2d37c9c522b07d4993c2eed19e into 300f5b403a6fd75cad6db5da42eea716ff0fa513 - [view on LGTM.com](https://lgtm.com/projects/g/zaproxy/zaproxy/rev/pr-babbd5361290d8704f13d31c0e21f535fff5b7ca)

**new alerts:**

* 3 for Unused local variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

psiinon commented on docker/zap_common.py:
```diff
@@ -47,8 +48,86 @@
+zap_hooks = None
+
+def load_custom_hooks(hooks_file=None):
+    """ Loads a your custom python module which modifies zap scripts behaviour
```
Either ' a ' or ' your ' not needed?

psiinon commented on docker/tests/test_zap_hooks.py:
```diff
@@ -0,0 +1,83 @@
+class TestZapHooks(unittest.TestCase):
+
+    @contextmanager
+    def assertNotRaises(self, exc_type=Exception):
```
Raised instead of Raises?

rezen replied:
I was trying to be consistent (and maybe illogical?) with the language of  `assertRaises` from stdlib but am not opposed to changing it.

thc202 commented on docker/tests/test_zap_hooks.py:
```diff
@@ -0,0 +1,83 @@
+
+
+    def test_trigger_hook_mismatch_no_exception(self):
+        """ If the hook signature doesn't match the hook, no exception is raised """
```
Why not? This seems to evidence an error in the provided hook.

thc202 commented on docker/tests/test_zap_hooks.py:
```diff
@@ -0,0 +1,83 @@
+
+
+    def test_trigger_hook_ignore_exception(self):
+        """ Hooks that raise exceptions are ignored when triggered """
```
This does not seem to be the best behaviour IMO, by continuing issues in the hooks could easily go unnoticed and the script would not be doing what the user expects/wants.

thc202 commented on docker/tests/test_zap_hooks.py:
```diff
@@ -0,0 +1,83 @@
+
+
+    def test_trigger_hook_verify_calls(self):
+        """ Should return same args if no hook to trigger """
```
Does not seem to match the test.

thc202 commented on docker/tests/test_zap_hooks.py:
```diff
@@ -0,0 +1,83 @@
+
+
+    def test_trigger_hook_maintain_signature(self):
+        """Should return same args if no hook to trigger """
```
Same here.

thc202 replied:
Bump, the description does not match the test (I think).

thc202 commented on docker/zap-api-scan.py:
```diff
@@ -148,12 +150,12 @@ def main(argv):
         usage()
         sys.exit(3)
-
+    
```
These trailing spaces should be removed (same in other places).

thc202 commented on docker/zap-api-scan.py:
```diff
@@ -206,8 +208,13 @@ def main(argv):
     check_zap_client_version()
+    
+    load_custom_hooks(hook_file)
+    opts = trigger_hook('cli_opts', opts)
```
What is this one for? Or, does it worth assigning the return value even if `opts` is never used after?

thc202 commented on docker/zap_common.py:
```diff
@@ -214,6 +294,7 @@ def wait_for_zap_start(zap, timeout_in_secs = 600):
           'Failed to connect to ZAP after {0} seconds'.format(timeout_in_secs))
 
 
+@hook(wrap=True)
```
What's the criteria being used to wrap?

thc202 replied:
Thanks.

thc202 commented on docker/zap_common.py:
```diff
@@ -423,3 +509,12 @@ def write_report(file_path, report):
+
+@hook(wrap=True)
+def zap_import_context(zap, context_file):
+    # handle the context file, cant use base_dir as it might not have been set up
```
This should be removed, does not apply here.

rezen replied:
Why do you say that?

thc202 replied:
There's no `base_dir` here, the comment was from callers' scope.

rezen replied:
Ah gotcha yes!

thc202 commented:
The command line help is not enough to use this feature, the above description and example usage should be added to a wiki page and then linked from the existing script ones (e.g. https://github.com/zaproxy/zaproxy/wiki/ZAP-Baseline-Scan).

psiinon commented on docker/tests/README.md:
```diff
@@ -0,0 +1,26 @@
+# Testing
```
Can we automatically run these tests in CI?

rezen commented:
@thc202 I wasn't sure how to update the wiki via a PR :/ should I just include a doc in this PR than can be copied into the wiki?

thc202 commented:
Yeah, the wiki does not support PRs, you can push directly though. Adding the doc here might be better, easier for review/feedback.

rezen commented:
Addressed comments in last push :smile:

thc202 commented on docker/docs/scan-hooks.md:
```diff
@@ -0,0 +1,71 @@
+before the rest of the scan runs.
+
+**Changing Configs**  
+For some applications, there are a number active scan scanners that you may want
```
`number of`?

thc202 commented on docker/docs/scan-hooks.md:
```diff
@@ -0,0 +1,71 @@
+## Example
+#### Define your hooks in a python file `my-hooks.py`
+You define all the hooks you want to integrate with with python methods that
+correspond with the name of the hook. By default, zap scans will look for 
```
`ZAP`.

thc202 commented on docker/tests/test_zap_hooks.py:
```diff
@@ -0,0 +1,62 @@
+
+
+    def test_trigger_hook_verify_calls(self):
+        """ Verify the hook gets called if it matchesShould return same args if no hook to triggerr signature """
```
This needs tweaks.

thc202 commented on docker/tests/test_zap_hooks.py:
```diff
@@ -0,0 +1,62 @@
+        """ Verify the hook gets called if it matchesShould return same args if no hook to triggerr signature """
+        zap_common.zap_hooks = _MockHooks()
+        args = ['zap', 'http://127.0.0.1']
+        zap_common.trigger_hook('zap_started', *args)Should return same args if no hook to trigger
```
Here (and following) as well?

rezen commented:
Reread the docs, fixed some additional grammatical errors. For the tests there was a haphazard paste that multiplied itself :disappointed: fixed that!

thc202 commented:
Thank you!

rezen commented:
If you like I can do a blog post on using this feature on my site?

thc202 commented:
Sounds great! It could also be published in ZAP's blog?

psiinon commented:
@rezen agreed, would be very happy for you to publish this on the ZAP blog :)

rezen commented:
Ya definitely please use the content! I'm a couple weeks weeks out from getting to it. I'll also use some examples to address some users' issues. The hook system is great for addressing one off features/requests to scripts :smile: 

- https://github.com/zaproxy/zaproxy/issues/3809
- https://github.com/zaproxy/zaproxy/issues/4354
